keps/sig-node/753-sidecar-containers/README.md

```diff
@@ -843,24 +843,117 @@ examples:
+as much as possible (think of a Spot Instances with 30 seconds notification).
+
+Here is the proposed approach:
+1. Sidecar containers that have a `PreStop` hook will be notified when the Pod has begun terminating
+   by executing the `PreStop` hook. This happens at the same time as regular containers, and begins
+   the Pod's termination grace period countdown.
+2. Once the last primary container terminates, the last started sidecar container is notified by
+   sending a `SIGTERM` signal.
+3. The next sidecar (in reverse order) is notified by sending a `SIGTERM` signal after the previous
+   sidecar container terminates.
+4. This continues until all sidecar containers have terminated, or the Pod's termination grace period
+   expires.
+5. In the latter case, all remaining containers are notified by a `SIGTERM`, followed by a fixed
+   grace period of 2s and finally terminated.
+6. The Pod will be terminated after that.
+
+Pseudocode for the above:
+
+```
+func terminatePod() {
+  // notify all sidecar containers with preStop hook, asynchronously
+  for sidecar in sidecarContainers {
+    if sidecar has preStop hook {
+      go execute preStop hook // async
+    }
+  }
+  // notify all containers with preStop hook and then SIGTERM, asynchronously
+  for container in containers {
+    if container has preStop hook {
+      go func(container) { // async
+        execute preStop hook
+        send SIGTERM
+      }
+    }
+  }
+  for {
+    switch {
+      case grace period expired:
+        for anyContainer in sidecarContainers + containers {
+          if anyContainer is running {
+            send SIGTERM
+          }
+        }
+        sleep 2s
+        for anyContainer in sidecarContainer + containers {
+          if anyContainer is running {
+            send SIGKILL
+          }
+        }
+        return
+      case all containers are terminated:
+        // sidecars are terminated in reverse order
+        for sidecar in reverse(sidecarContainers) {
+          // sidecar is already terminating, let it finish
+          if sidecar is terminating {
+            break
+          }
+          // next sidecar to terminate
+          else if sidecar is running {
+            send SIGTERM
+            break
+          }
+        }
+        sleep 1s
+      case all sidecarContainers are terminated:
+        return
+      default:
+        sleep 1s
+    }
+  }
+}
+```
 
+It is worth noting that, like with regular containers, `PreStop` hook must complete before the `SIGTERM`
+signal to stop the sidecar container can be sent. Therefore, ordering and graceful termination of sidecars
+can only be guaranteed if the `PreStop` hook completes within the Pod's termination grace period.
+
+Sidecars continue to be restarted until they enter the `Terminated` state which they are notified
+by a `SIGTERM` signal. This is to ensure that sidecars that fail are restarted until the TGPS expires.
```

Comment:
Also say that the `preStop` will be called on restarted container

Reply:
yes it was implied